### PR TITLE
Signing data draft post

### PR DESCRIPTION
**Pull request by GuiBibeau**
Post flows like this:

1. Introduce signing and what it means
2. give a little history
3. technical overview
4. conslusion and tips

Missing:

- [ ] Link to some EIPs and docs
- [ ] refactor our own docs to use JSON RPC instead of web3 (in current MetaMask docs, we show examples with web3.js)

### CODE REVIEW

File: signing-with-metamask/index.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+It's essential to keep in mind the security and user experience, as well as to validate the data and use a secure connection to prevent man-in-the-middle attacks. By following best practices and implementing the appropriate method, developers can ensure the integrity and authenticity of the data signed using MetaMask. It's also important to note that, to use these methods, the user must have MetaMask installed and configured on their browser and have granted the dApp permission to access their MetaMask account.
+
+MetaMask provides several methods for signing data and transactions, each method has its own specific use case, and developers should choose the appropriate method for their use case. To ensure the security and integrity of the data, developers should also follow best practices. The MetaMask team highly encourages the use of `signedTypeDataV4` as a default and revert to `personal_sing` in edge cases. Remember to implement appropriate measures to validate the data, use a secure connection and test the code thoroughly before deploying it to production.
```
Comment (httpJunkie) [inline signing-with-metamask/index.md]: Could we update the conclusion to have our recommendation at the top, followed by a more brief conclusion and support for that recommendation?

Also, is there any plan to include an example of using the three methods?

File: signing-with-metamask/index.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+2019: MetaMask introduced the ability to sign and send transactions directly from the browser extension, eliminating the need for users to copy and paste the transaction data into a separate signing tool.
+2020: MetaMask introduced the ability to sign transactions using hardware wallets, such as Trezor and Ledger, providing users with an additional layer of security.
+2021: MetaMask introduced the ability to sign messages using a new, more secure method called the "Ethereum Message Signing Protocol (EMSP)".
+
+MetaMask has continued to evolve and improve throughout its journey, providing users with more secure and user-friendly ways to sign data and transactions on the Ethereum blockchain. With the latest updates, MetaMask is now one of the most popular tools for signing data and transactions on the Ethereum network, used by developers and users alike.
+
+Using ethers.js for signing data and transactions on the Ethereum blockchain with MetaMask is convenient, providing an alternative approach. It offers a developer-friendly API and better support for TypeScript, which makes it easier for developers to build dApps and interact with the Ethereum blockchain.
```
Comment (httpJunkie) [inline signing-with-metamask/index.md]: The last 3 paragraphs in this section feel like we are reiterating some of the same information can we condense this down to one or two paragraphs that ensure no redundancy?

**Review by httpJunkie — CHANGES_REQUESTED** [review]

File: signing-with-metamask/index.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+When you sign data with MetaMask, your private key never leaves your device, so your key is safe from hackers and other malicious actors who might try to steal it. MetaMask uses a technique known as "hashing" to ensure that your data cannot be tampered with and is secure.
+
+Below are some of the critical milestones in the journey of signing data and transactions with MetaMask:
+
+2016: MetaMask was first released as a browser extension for Google Chrome, allowing users to interact with the Ethereum blockchain and sign transactions with their private keys.
+2017: MetaMask added support for signing data and messages, allowing users to prove ownership of an Ethereum address.
+2018: MetaMask introduced the ability to sign typed data, a specific format for structured data specified by the EIP-712 standard. This feature enabled users to sign more structured and organized data.
+2019: MetaMask introduced the ability to sign and send transactions directly from the browser extension, eliminating the need for users to copy and paste the transaction data into a separate signing tool.
```
Comment (Gudahtt) [inline signing-with-metamask/index.md]: I'm not sure what this refers to. I believe MetaMask has had the ability to send transactions since the beginning, or very close to it.

File: signing-with-metamask/index.md [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+2020: MetaMask introduced the ability to sign transactions using hardware wallets, such as Trezor and Ledger, providing users with an additional layer of security.
```
Comment (Gudahtt) [inline signing-with-metamask/index.md]: Ledger support was added in 2018: https://github.com/MetaMask/metamask-extension/blob/develop/CHANGELOG.md#4100---2018-09-18
As was Trezor support: https://github.com/MetaMask/metamask-extension/blob/develop/CHANGELOG.md#490---2018-08-07

File: signing-with-metamask/index.md [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+2021: MetaMask introduced the ability to sign messages using a new, more secure method called the "Ethereum Message Signing Protocol (EMSP)".
```
Comment (Gudahtt) [inline signing-with-metamask/index.md]: Hmm, I've never heard of this. Do you have any more information about this?

File: signing-with-metamask/index.md (hunk 2 above)
Comment (Gudahtt) [reply signing-with-metamask/index.md]: This discussion of Ethers.js feels out of place here as well. It doesn't seem especially relevant for signing. Where the convenience libraries really shine are interactions with contracts, which are quite complex without them. But signing is simple.

The part about types is fair - for a TypeScript project, you might want to use a convenience library that is typed. But this won't be a factor for very long, as we're actively working on providing types for our RPC API. And the provider API is already typed (here: https://github.com/MetaMask/providers/blob/main/src/MetaMaskInpageProvider.ts), though I'm not aware of any docs that explain how to use them.

File: signing-with-metamask/index.md [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+While MetaMask can provide a user-friendly interface for signing data and transactions, ethers.js, an alternative to web3.js (another JavaScript library for interacting with the Ethereum blockchain), has been focused on providing a developer-friendly library for building dApps and interacting with the Ethereum blockchain. Developers have widely adopted ethers.js as a more intuitive and easy-to-use library for signing data and transactions.
+
+MetaMask and ethers.js are not mutually exclusive and can be used together; developers can use MetaMask to interact with the Ethereum blockchain and sign data and transactions, and use ethers.js to build their dApp's front-end. ethers.js may also be used to sign data and transactions on other Ethereum clients such as Geth or Parity.
+
+## Methods of Signing Data With MetaMask
+
+We'll explore various ways of signing data using the provider injected by MetaMask as well as implementations in web3.js and ethers.js:
+
+eth_sign: This method is used to sign data with the Ethereum address's private key and returns the signature in the format of `"0x<r><s><v>"`.
```
Comment (Gudahtt) [inline signing-with-metamask/index.md]: It would be better to de-emphasize our `eth_sign` method as much as possible. It has been deprecated for years, and we're considering removing it by default in an upcoming release. It poses a substantial risk to users. If we're to mention it at all, it should be last.

File: signing-with-metamask/index.md [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+The `eth_sign` method is used to sign data and prove ownership of an Ethereum address. A limitation of this method is that it does not include any prefix before signing the data, which could lead to the signature not being recognized as valid by some smart contracts. It also poses a security risk, so it should be avoided.
+
+personal_sign: This method is also used to sign data and prove ownership of an Ethereum address, similar to `eth_sign,` but adds an "Ethereum Signed Message" prefix to the data before signing it. It is part of the JSON-RPC personal module, which is for personal account management. It signs data with a user's private key and returns the signature.
```
Comment (Gudahtt) [inline signing-with-metamask/index.md]: This might be confusing for developers that are also working with other wallets, because `personal_sign` isn't just similar to `eth_sign` - it _is_ `eth_sign`, or at least the updated standardized form of it.

Our `eth_sign` follows an older specification for that signing method. It was updated to use the "Ethereum Signed Message" prefix, but we left our implementation without that prefix to avoid breaking dapps that relied on it. 

Instead we adopted `personal_sign` but implemented it like the modern `eth_sign`. This is confusing as well, because `personal_sign` has a spec that we're not following as well! It's specified to be identical to the modern `eth_sign` except that it also accepts a password for the account, but we don't accept any password.

So someone coming from a different wallet might:

* Expect that `eth_sign` includes a prefix, but ours does not
* Expect `eth_sign` to be the standard way of signing a plain string, but for us it's `personal_sign`,
* Expect `personal_sign` to accept a password, but it does not

This will probably be tricky to explain. Not exactly sure how to approach it, or how much detail to include. For a new dev, fewer details would be better surely. I just wanted to highlight this risk of confusion.

File: signing-with-metamask/index.md [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+personal_sign: This method is also used to sign data and prove ownership of an Ethereum address, similar to `eth_sign,` but adds an "Ethereum Signed Message" prefix to the data before signing it. It is part of the JSON-RPC personal module, which is for personal account management. It signs data with a user's private key and returns the signature. 
+
+"`tsx
+window.ethereum.request({ method: 'personal_sign',  params: [msg, from],  });
+```
+
+Where `data` is the data to be signed, `address` is the address of the user signing the data and `callback` is an optional callback function to handle the result. Advantages of this method include the added security provided by the prefix, which allows the signature to be recognized as valid by more smart contracts. However, a limitation of this method is that it only supports UTF-8 encoded data. Developers should choose this method if they need to sign data that will be recognized as valid by more smart contracts, and the data is UTF-8 encoded.
```
Comment (Gudahtt) [inline signing-with-metamask/index.md:61]: Hmm. It seems a bit confusing to say that the data is UTF-8 encoded, when we accept it as a JavaScript string. Strings are UTF-16 encoded. Talking about encoding would make sense if we were accepting raw bytes, but we're not.

Instead maybe it would be better to say that this method only supports string messages. That's the limitation, that it's a string. That would be easier to understand and less misleading.

Comment (Gudahtt) [inline signing-with-metamask/index.md]: Hmm. The encoding is not the major limitation here, it's that the message has to be a string. Mentioning the encoding is fine, but we should call out that it's a string. That's the part that might be a restriction in practice. I don't think anybody cares about the string encoding; UTF-8 is the standard string encoding used nearly everywhere.

File: signing-with-metamask/index.md [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+
+
+eth_signTypedDataV4: This method is used to sign typed data (a specific format for structured data specified by the EIP-712 standard) and prove ownership of an Ethereum address, and returns the signature in the format of `0x<r><s><v>`.
+
+"`tsx
+window.ethereum.request({ method: 'eth_signTypedData_v4', params: [address, data] })
+```
+
+Where `data` is the typed data object to be signed, `address` is that of the user signing the data, and `callback` is an optional callback function to handle the result. Advantages of this method include its support for typed data, which allows for more structured and organized data, and its support for the EIP-712 standard, which is widely recognized and supported by the Ethereum community. However, a limitation of this method is that it only supports typed data, which means it can't be used to sign just any type of data. Developers should choose this method if they need to sign typed data according to the [EIP-712](https://eips.ethereum.org/EIPS/eip-712) standard.
```
Comment (Gudahtt) [inline signing-with-metamask/index.md]: >Developers should choose this method if they need to sign typed data according to the [EIP-712]

This makes it sound like you probably don't want to use this signature type if you're unfamiliar with the EIP. That doesn't seem like good advice.

For comparison, check out the updated signing docs that Dan Finlay recently wrote: https://docs.metamask.io/guide/signing-data.html

It suggests using this method when you want readable signatures that are easily processed on-chain. This is a good description.

Personally I'd suggest using `personal_sign` for plain text, but using `eth_signTypedData_v4` for structured data.

Comment (Gudahtt) [inline signing-with-metamask/index.md]: >a limitation of this method is that it only supports typed data, which means it can't be used to sign just any type of data

This sentence is a bit confusing. A reader might get the impression that if their data isn't already "typed", then it can't be used here. But in the typical case, the user of this method would be the ones writing/generating these types.

Maybe it would be better to say something along the lines of.... this might be unsuitable or less useful for data that can't be described easily with types? Something along those lines.

File: signing-with-metamask/index.md [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+When using these methods, it's important to remember that the user's private key must be kept safe and secure. It's also important to validate the data before signing it and to use a secure connection to prevent man-in-the-middle attacks.
```
Comment (Gudahtt) [inline signing-with-metamask/index.md]: Hmm, it's not entirely clear what the point was re: the user's private key. How does that relate to signing, and what is the reader expected to take away from that advice?

And for validating the data, was this meant for users? Most of this article is directed at devs. So maybe for devs, the recommendation would be to use signatures that are easily to validate, or something like that.

I'd recommending checking out our new signing docs, they have good advice for devs on security and usability.

File: signing-with-metamask/index.md (hunk 1 above)
Comment (Gudahtt) [inline signing-with-metamask/index.md]: Typo: `personal_sing` => `personal_sign`

**Comment by GuiBibeau** [comment]
@Gudahtt @httpJunkie Looking at this, I'm confident I missed the mark and I want to draft a rewrite with your comments in mind. I'll re push a big change by end of day.

**Review by httpJunkie — APPROVED** [review]
LGTM
